cpp/src/arrow/extension/variable_shape_tensor.h

```diff
@@ -0,0 +1,110 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include "arrow/extension_type.h"
+
+namespace arrow {
+namespace extension {
+
+class ARROW_EXPORT VariableShapeTensorArray : public ExtensionArray {
+ public:
+  using ExtensionArray::ExtensionArray;
+};
+
+/// \brief Concrete type class for variable-shape Tensor data.
+/// This is a canonical arrow extension type.
+/// See: https://arrow.apache.org/docs/format/CanonicalExtensions.html
+class ARROW_EXPORT VariableShapeTensorType : public ExtensionType {
+ public:
+  VariableShapeTensorType(const std::shared_ptr<DataType>& value_type, const int32_t ndim,
+                          const std::vector<int64_t> permutation = {},
+                          const std::vector<std::string> dim_names = {},
+                          const std::vector<std::optional<int64_t>> uniform_shape = {})
+      : ExtensionType(struct_({::arrow::field("shape", fixed_size_list(int32(), ndim)),
+                               ::arrow::field("data", list(value_type))})),
```

Comment:
The spec puts the `data` field before the `shape` field:
https://arrow.apache.org/docs/format/CanonicalExtensions.html#variable-shape-tensor

Reply:
Nice catch. Changed to match spec.